src/vsr/replica.zig

```diff
@@ -1400,46 +1443,105 @@ pub fn ReplicaType(
-                @panic("unimplemented (state transfer)");
-            }
-
             assert(self.status == .view_change or
                 self.status == .normal or
                 self.status == .recovering_head);
             assert(message.header.view >= self.view);
             assert(message.header.replica != self.replica);
             assert(message.header.replica == self.primary_index(message.header.view));
+            assert(message.header.op >= message.header.commit);
+            assert(message.header.op - message.header.commit <=
+                constants.pipeline_prepare_queue_max);
+
+            if (message.header.view == self.log_view and message.header.op <= self.op) {
+                // We were already in this view prior to receiving the SV.
+                assert(self.status == .normal or self.status == .recovering_head);
+
+                log.debug("{}: on_start_view view={} (ignoring, old message)", .{
+                    self.replica,
+                    self.log_view,
+                });
+                return;
+            }
 
             self.view_jump(message.header);
 
-            assert(self.status == .view_change);
+            assert(self.status == .view_change or self.status == .normal);
+            assert(self.status == .view_change or self.backup());
+            assert(self.status == .view_change or self.log_view == self.view);
             assert(message.header.view == self.view);
-            assert(message.header.op == op_highest(message_body_as_headers(message)));
 
-            self.set_op_and_commit_max(message.header.op, message.header.commit, "on_start_view");
-            for (message_body_as_headers_chain_consecutive(message)) |*header| {
-                self.replace_header(header);
+            self.view_headers = vsr.Headers.ViewChangeArray.from_slice(message_body_as_headers_chain_disjoint(message));
+            assert(self.view_headers.array.get(0).view <= self.view);
+            assert(self.view_headers.array.get(0).op == message.header.op);
+
+            {
+                const primary_repair_min =
+                    self.view_headers.array.get(self.view_headers.array.len - 1).op;
+                var op = self.commit_min + 1;
+
+                // Use "op ≤ primary_repair_min" instead of "=" to force the logs to intersect.
+                // This will cause a state transfer in the case where:
+                // - the repairing replica has a pristine log up to its op_checkpoint_trigger=X
+                // - the SV headers begin at op X+1
+                // But in exchange it avoids a special-case to update the commit_max and verify
+                // X's header.
+                const repairable = while (op <= primary_repair_min) : (op += 1) {
+                    const slot = self.journal.slot_with_op(op) orelse break false;
+                    if (self.journal.dirty.bit(slot)) break false;
+                } else true;
```

Comment:
We are not using `find_latest_headers_break_between` here, because that only checks headers, and here we care about prepares, right? That is, `self.journal.dirty.bit(slot)` tell us whether we have a prepare, right? 

Maybe add this as helper fn in Journal, right next to `find_latest_headers_break_between` (unless I misunderstand the semantics, and the two are unrelated)?

Also, is `self.journal.dirty.bit(slot)` the right condition to use here? I _think_ the dirty bit is set, while we are writing a prepare. But that seems fine? If we have prepare somewhere in memory (in cache?), we shouldn't try to repair it?